DIY ngrok/pagekite for Android app testing #324 (#345)
medic/cht-docs: content/en/apps/guides/debugging/Securely-sharing-your-development-environment.md

---
title: "Securely sharing your development environment"

> linkTitle: "Securely sharing your development environment"

Reviewer: The title is clearer now, although it should be in title case. 

Also, I would suggest making the filename more formal, since it will likely live on even if the title gets tweaks, eg `secure-sharing-of-developer-instance`

Author: Cool cool - should be fixed in 609baf2

weight: 
description: >
  Use a publicly accessible Linux web server to forward https requests to your development environment
relatedContent: >
  
---

## Overview
When using a local [development environment](https://github.com/medic/cht-core/blob/master/DEVELOPMENT.md), you may want to share your work with other collaborators. You also may need to access the environment from your [mobile device](https://github.com/medic/medic-android/) which requires an SSL certificate (the "s" in "https"). By using a publicly accessible web server, you can receive the secure https requests and forward them back to your CHT instance which doesn't have https set up:

[<img src="/apps/guides/debugging/images/SSH.tunnel.diagram.svg" width=100% height=100%>](/apps/guides/debugging/images/SSH.tunnel.diagram.svg)

Once you have this web server set up, you many continue to use when ever you want by simply reconnecting to it via the secure tunnel.

## Prereqs

> This guide assume:

Reviewer 2: ```suggestion
This guide assumes:
```

Author: add missing "s" to "assume"

> * You have a local dev instances set up of cht-core https://github.com/medic/cht-core/blob/master/DEVELOPMENT.md

Reviewer 2: ```suggestion
* You have a local dev instance set up of cht-core https://github.com/medic/cht-core/blob/master/DEVELOPMENT.md
```

* You have medic-android repo checked out and can compile the android app from scratch https://github.com/medic/medic-android/
* You  have the generic Medic Mobile app installed on your android device https://play.google.com/store/apps/details?id=org.medicmobile.webapp.mobile&hl=en_US

> * You have an Ubuntu >18.04 server with a public IP and a DNS entry that you can SSH into and have sudo on

Reviewer 2: Would '>=' to be more precise here?

Author: Hi @andy5995! Thanks so much for your feedback on this PR.

I accepted your others, but this one felt a little awkward.  Checking other instances of versions in this repo, they do "use X.X.X or later".  I did a light refactor of this sentence to do the same in 609baf2

Reviewer 2: You're welcome @mrjones-plip!

* You have apache >2.4.29 installed on the Ubuntu server and can add a new vhost to it, including an SSL cert. (nginx could be used instead as well, but not covered here)
* You have certbot installed from letsencrypt.org

> For reference any of the cheap servers out there (Digital Ocean has a $5/mo server https://digitalocean.com/) will enable you to do this. 

Reviewer 2: ```suggestion
For reference, any of the cheap servers out there (Digital Ocean has a $5/mo server https://digitalocean.com/) will enable you to do this. 
```

> ## Warning!!1!
> Be extra careful with this process! The end result will be that your development instance will be accessible to the internet. If you have simple logins and passwords like "admin/test.223" because you thought it was just your local dev instance and it doesn't matter, now it matters! Whenever you're not using the SSH tunnel for testing, shut it down so not remote access is allowed.
> 
> Never expose a development instance to the internet where you've replicated production data locally. Well, maybe not never, but with extreme care and intention.

Reviewer: Would be good to have this warning at the top of the page. Here are some suggested edits to match the tone of the docs site, along with the use of an "alert" box.

```suggestion
{{% alert title="Warning" %}} 
Be extra careful with this process as your development instance and its data will be accessible to the internet. If you have simple or default logins and passwords like "admin/password" because you thought it was just your local dev instance and it doesn't matter, now it matters! 

Only follow this procedure if you are certain that you are not exposing production data that may have protected health information. Whenever you're not using the SSH tunnel for testing, shut it down so remote access not is allowed.
{{% /alert %}}

```

Author: Good thinking! Reformatted and moved to the top per 609baf2
